docs/datamodel/build-a-datamodel.rst

```diff
@@ -0,0 +1,627 @@
+
+Note, that the left-hand-side of the entry point
+``records = my_datamodel.mappings``, defines the folder name/index alias (i.e.
+``records`` and that the right-hand-side defines the Python import path to the
+``mappings`` package.
+
+.. note::
+
+    A typical mistake is to forget to add a blank ``__init__.py`` file inside
+    the ``mappings``, ``v5`` and ``v6`` folders, in which case the entry points won't be correctly discovered.
+
+Define a Marshmallow schema
+---------------------------
+`Marhsmallow <https://marshmallow.readthedocs.io/en/3.0/index.html>`_ is a
+Python library that helps you write highly advanced
+serialization/deserialization/validation rules for your input/output data.
+
+Marshmallow use in Invenio is optional, but is usually very helpful when you go
+beyond purely structural data validation - e.g. validating one field given the
+value of another field.
+
+In Invenio the Marshmallow schemas are located in the ``marshmallow`` Python
+module. You may have multiple Marshmallow schemas depending on your
+serialization and deserialization needs.
+
+.. code-block:: shell
+
+    |-- my_datamodel
+    |   |-- marshmallow
+    |   |   |-- __init__.py
+    |   |   `-- json.py
+
+Below is a simplified example of a Marshmallow schema you could use in
+``json.py`` (note, the scaffolded data model package, includes a more complete
+example):
+
+.. code-block:: python
+
+    from invenio_records_rest.schemas import StrictKeysMixin
+    from marshmallow import fields
+
+    class RecordSchemaV1(StrictKeysMixin):
+        metadata = fields.Raw()
+        created = fields.Str()
+        revision = fields.Integer()
+        updated = fields.Str()
+        links = fields.Dict()
+        id = fields.Str()
+
+In Invenio the Marshmallow schemas are often used together with serializers and
+loaders, so continue reading to see how the schema is used.
+
+**What's the difference: JSONSchemas, Mappings and Marshmallow?**
+
+It may seem a bit confusing that Invenio is dealing with three types of
+schemas. There's however good reasons:
+
+- **JSONSchema**: Deals with the internal structural validation of records
+  stored in the database (much like you define the table structure in
+  database).
+- **Elasticsearch mappings**: Deals with how records are indexed in
+  Elasticsearch which has big impact on your search results.
+- **Marshmallow schema**: Deals with primarily data validation and
+  transformation for both serialization and deserialization.
+
+Define serializers
+------------------
+Think of serializers as the definition of your output formats for records. The
+serializers are responsible for transforming the internal JSON for a record
+into some external representation (e.g. another JSON format or XML).
+
+Serializers are defined in the ``serializers`` module:
+
+.. code-block:: shell
+
+    |-- my_datamodel
+    |   |-- serializers
+    |   |   `-- __init__.py
+
+By default, Invenio provides serializers that can help you serialize your
```

Comment:
Since in all previous examples, when a new concept is introduced, there is a partial directory tree to show where the concept is situated and then a simple code snippet of the concept itself, a JSONSchema, ES mapping etc... Would it make sense here to link to the mentioned [serializers](https://github.com/inveniosoftware/invenio-records-rest/blob/master/invenio_records_rest/serializers) so, we mention that they are provided by default and also for the curious/advanced reader, we provide them with the location of the files?

I understand the risk since these files are not trivial as the previous snippets but I am sure that when you read that something is already provided and you are not super interested on it, you will just ignore the link.